src/main/java/dan200/computercraft/client/render/TileEntityMonitorRenderer.java

```diff
@@ -191,28 +191,27 @@ private static void renderTerminal( @Nonnull MultiBufferSource bufferSource, Pos
                 var vbo = monitor.buffer;
                 if( redraw )
                 {
-                    int vertexSize = RenderTypes.MONITOR.format().getVertexSize();
-
                     // Draw the background contents of the terminal into one vbo.
-                    ByteBuffer buffer = getBuffer( DirectFixedWidthFontRenderer.getVertexCount( terminal ) * vertexSize );
-
-                    DirectFixedWidthFontRenderer.drawTerminalBackground(
-                        buffer, 0, 0, terminal, !monitor.isColour(), yMargin, yMargin, xMargin, xMargin
+                    var sink = IrisCompat.INSTANCE.getVertexSink(
+                        DirectFixedWidthFontRenderer.getVertexCount( terminal ),
+                        TileEntityMonitorRenderer::getBuffer
                     );
+                    int vertexSize = sink.getFormat().getVertexSize();
+                    var buffer = sink.buffer();
+
+                    DirectFixedWidthFontRenderer.drawTerminalBackground( sink, 0, 0, terminal, !monitor.isColour(), yMargin, yMargin, xMargin, xMargin );
                     int backgroundBytes = buffer.position();
-                    DirectFixedWidthFontRenderer.drawTerminalForeground(
-                        buffer, 0, 0, terminal, !monitor.isColour(), yMargin, yMargin, xMargin, xMargin
-                    );
+                    DirectFixedWidthFontRenderer.drawTerminalForeground( sink, 0, 0, terminal, !monitor.isColour(), yMargin, yMargin, xMargin, xMargin );
                     int termVertices = buffer.position() / vertexSize;
                     monitor.backgroundVertexCount = backgroundBytes / vertexSize;
                     monitor.foregroundVertexCount = termVertices - monitor.backgroundVertexCount;
 
                     // If the cursor is visible, we append it to the end of our buffer. When rendering, we can either
                     // render n or n+1 quads and so toggle the cursor on and off.
-                    DirectFixedWidthFontRenderer.drawCursor( buffer, 0, 0, terminal, !monitor.isColour() );
+                    DirectFixedWidthFontRenderer.drawCursor( sink, 0, 0, terminal, !monitor.isColour() );
 
                     buffer.flip();
-                    vbo.upload( termVertices, RenderTypes.MONITOR.mode(), RenderTypes.MONITOR.format(), buffer );
+                    vbo.upload( buffer.position() / vertexSize, RenderTypes.MONITOR.mode(), sink.getFormat(), buffer );
```

Comment:
`buffer.flip()` should be moved to after the `vbo.upload( buffer.position() / ...` line. It's passing 0 as the vertex count now. Since we use the vertex count to ensure storage in the index buffer, we end up overrunning the buffer on complex monitors if nobody else has grown it out.

This is the source of this type of error that you showed me on Discord:
![2022-06-07_17 09 46](https://user-images.githubusercontent.com/748280/172516107-f0563616-ecef-43f6-b5c4-e67433321e01.png)

Also, this is my fault! I knew we should be including the cursor vertices in the count, and I remember trying to do it this same way. I couldn't figure it out so I left it for somebody else to trip over :P

Reply:
Well I was wrong again: we have to flip the buffer before passing to `vbo.upload`. I guess we need to calculate the total vertices before the flip. I'd like to clean up some of the other mess I left in this section too:

```patch
--- a/src/main/java/dan200/computercraft/client/render/TileEntityMonitorRenderer.java
+++ b/src/main/java/dan200/computercraft/client/render/TileEntityMonitorRenderer.java
@@ -200,18 +200,17 @@ private static void renderTerminal( @Nonnull MultiBufferSource bufferSource, Pos
                     var buffer = sink.buffer();
 
                     DirectFixedWidthFontRenderer.drawTerminalBackground( sink, 0, 0, terminal, !monitor.isColour(), yMargin, yMargin, xMargin, xMargin );
-                    int backgroundBytes = buffer.position();
+                    monitor.backgroundVertexCount = buffer.position() / vertexSize;
                     DirectFixedWidthFontRenderer.drawTerminalForeground( sink, 0, 0, terminal, !monitor.isColour(), yMargin, yMargin, xMargin, xMargin );
-                    int termVertices = buffer.position() / vertexSize;
-                    monitor.backgroundVertexCount = backgroundBytes / vertexSize;
-                    monitor.foregroundVertexCount = termVertices - monitor.backgroundVertexCount;
+                    monitor.foregroundVertexCount = (buffer.position() / vertexSize) - monitor.backgroundVertexCount;
 
                     // If the cursor is visible, we append it to the end of our buffer. When rendering, we can either
                     // render n or n+1 quads and so toggle the cursor on and off.
                     DirectFixedWidthFontRenderer.drawCursor( sink, 0, 0, terminal, !monitor.isColour() );
 
+                    int totalVertices = buffer.position() / vertexSize;
                     buffer.flip();
-                    vbo.upload( buffer.position() / vertexSize, RenderTypes.MONITOR.mode(), sink.getFormat(), buffer );
+                    vbo.upload( totalVertices, RenderTypes.MONITOR.mode(), sink.getFormat(), buffer );
                 }
 
                 // TODO Figure out how to setup the inverse view rotation matrix properly.
```